File: test_prohandel_api.py
"""
Test script to directly test the ProHandel API endpoints
"""
import os
import json
import requests
from requests.exceptions import RequestException
import traceback

# Load environment variables from .env file
def load_env_vars():
    env_vars = {}
    try:
        with open('.env', 'r') as f:
            for line in f:
                if line.strip() and not line.startswith('#'):
                    key, value = line.strip().split('=', 1)
                    env_vars[key] = value
        return env_vars
    except Exception as e:
        print(f"Error loading .env file: {e}")
        return {}

# Get API credentials
env_vars = load_env_vars()
API_KEY = env_vars.get('PROHANDEL_API_KEY', '7e7c639358434c4fa215d4e3978739d0')
API_SECRET = env_vars.get('PROHANDEL_API_SECRET', '1cjnuux79d')
AUTH_URL = env_vars.get('PROHANDEL_AUTH_URL', 'https://auth.prohandel.cloud/api/v4')
API_URL = env_vars.get('PROHANDEL_API_URL', 'https://linde.prohandel.de/api/v2')

def authenticate():
    """Authenticate with the ProHandel API and get an access token"""
    print(f"Authenticating with {AUTH_URL}...")
    try:
        auth_data = {
            "apiKey": API_KEY,
            "secret": API_SECRET
        }
        
        response = requests.post(f"{AUTH_URL}/token", json=auth_data)
        response.raise_for_status()
        
        auth_response = response.json()
        access_token = auth_response["token"]["token"]["value"]
        print("Authentication successful!")
        return access_token
    except Exception as e:
        print(f"Authentication failed: {str(e)}")
        traceback.print_exc()
        return None

def test_endpoint(endpoint, access_token, params=None):
    """Test a specific API endpoint"""
    if params is None:
        params = {"page": 1, "pagesize": 10}
    
    # List of URL patterns to try, based on the working implementation
    url_patterns = [
        f"{API_URL}/{endpoint.lstrip('/')}",                  # Standard pattern
        f"{API_URL}/v2/{endpoint.lstrip('/')}",              # With v2 prefix
        f"{API_URL}/api/v2/{endpoint.lstrip('/')}",          # With api/v2 prefix
        f"{API_URL}/v2.29.1/{endpoint.lstrip('/')}",         # With version number
        f"https://linde.prohandel.de/api/v2/{endpoint.lstrip('/')}"  # Full URL with v2
    ]
    
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json",
        "Accept": "application/json",
        "X-API-Version": "v2.29.1"
    }
    
    print(f"\nTesting endpoint: {endpoint}")
    print(f"Parameters: {params}")
    
    for url in url_patterns:
        print(f"\nTrying URL: {url}")
        try:
            response = requests.get(url, headers=headers, params=params)
            print(f"Status code: {response.status_code}")
            
            if response.status_code == 200:
                data = response.json()
                
                # Check if response is a list or dictionary
                if isinstance(data, list):
                    print(f"Received list response with {len(data)} items")
                    if data:
                        print(f"First item sample: {json.dumps(data[0], indent=2)[:500]}...")
                elif isinstance(data, dict):
                    # Check if there's a data field
                    if 'data' in data:
                        items = data['data']
                        if isinstance(items, list):
                            print(f"Received {len(items)} items in 'data' field")
                            if items:
                                print(f"First item sample: {json.dumps(items[0], indent=2)[:500]}...")
                        else:
                            print(f"'data' field is not a list: {type(items)}")
                            print(f"Data sample: {json.dumps(data, indent=2)[:500]}...")
                    else:
                        print(f"No 'data' field in response")
                        print(f"Response keys: {list(data.keys())}")
                        print(f"Data sample: {json.dumps(data, indent=2)[:500]}...")
                else:
                    print(f"Unexpected response type: {type(data)}")
                
                # Return success status and data for the first successful URL
                return True, data
            else:
                print(f"Error response: {response.text}")
                # Continue to the next URL pattern
        except RequestException as e:
            print(f"Request error: {str(e)}")
            # Continue to the next URL pattern
        except Exception as e:
            print(f"Error: {str(e)}")
            # Continue to the next URL pattern
    
    # If all URL patterns failed, return failure
    print("All URL patterns failed for endpoint", endpoint)
    return False, None

def main():
    """Main function to test all endpoints"""
    # Authenticate
    access_token = authenticate()
    if not access_token:
        print("Authentication failed, cannot proceed with tests")
        return
    
    # List of endpoints to test
    endpoints = ["article", "customer", "order", "sale", "inventory"]
    
    # Test each endpoint
    results = {}
    for endpoint in endpoints:
        print(f"\n{'=' * 50}")
        print(f"Testing {endpoint.upper()} endpoint")
        print(f"{'=' * 50}")
        success, data = test_endpoint(endpoint, access_token)
        results[endpoint] = {
            "success": success,
            "has_data": bool(data and (isinstance(data, list) or 
                                      (isinstance(data, dict) and data.get('data'))))
        }
    
    # Print summary
    print("\n\n" + "=" * 50)
    print("TEST RESULTS SUMMARY")
    print("=" * 50)
    for endpoint, result in results.items():
        status = "✅ SUCCESS" if result["success"] else "❌ FAILED"
        has_data = "✅ HAS DATA" if result["has_data"] else "❌ NO DATA"
        print(f"{endpoint.upper()}: {status} | {has_data}")

if __name__ == "__main__":
    main()

File: scripts/test-prohandel-api.js
// Test script for ProHandel API integration
require('dotenv').config({ path: '.env.local' });
const fetch = require('node-fetch');

const PROHANDEL_CONFIG = {
  API_KEY: process.env.PROHANDEL_API_KEY,
  API_SECRET: process.env.PROHANDEL_API_SECRET,
  AUTH_URL: process.env.PROHANDEL_AUTH_URL,
  API_URL: process.env.PROHANDEL_API_URL,
  TENANT_ID: process.env.TENANT_ID,
};

// Authenticate and get token
async function authenticate() {
  try {
    console.log('Authenticating with ProHandel API...');
    const response = await fetch(`${PROHANDEL_CONFIG.AUTH_URL}/oauth/token`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        grant_type: 'client_credentials',
        client_id: PROHANDEL_CONFIG.API_KEY,
        client_secret: PROHANDEL_CONFIG.API_SECRET,
        tenant_id: PROHANDEL_CONFIG.TENANT_ID
      }),
    });

    if (!response.ok) {
      throw new Error(`Authentication failed: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    console.log('Authentication successful!');
    return data.access_token;
  } catch (error) {
    console.error('Authentication error:', error.message);
    process.exit(1);
  }
}

// Test getting warehouses/locations
async function testGetWarehouses(token) {
  try {
    console.log('\nTesting /locations endpoint...');
    const response = await fetch(`${PROHANDEL_CONFIG.API_URL}/locations`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'X-Tenant-ID': PROHANDEL_CONFIG.TENANT_ID,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch warehouses: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    console.log('Warehouse/Locations data:');
    console.log(JSON.stringify(data, null, 2));
    return data;
  } catch (error) {
    console.error('Error fetching warehouses:', error.message);
    return null;
  }
}

// Test getting stockout risk data with warehouse filter
async function testGetStockoutRiskByWarehouse(token, warehouseId) {
  try {
    console.log(`\nTesting /inventory/stockout-risks with warehouse ${warehouseId}...`);
    const response = await fetch(`${PROHANDEL_CONFIG.API_URL}/inventory/stockout-risks?location_id=${warehouseId}`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'X-Tenant-ID': PROHANDEL_CONFIG.TENANT_ID,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch stockout data: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    console.log('Stockout risk data for warehouse:');
    console.log(JSON.stringify(data.slice(0, 2), null, 2)); // Show just the first 2 items
    console.log(`...and ${data.length - 2} more items`);
    return data;
  } catch (error) {
    console.error('Error fetching stockout data for warehouse:', error.message);
    return null;
  }
}

// Test getting sales data with warehouse filter
async function testGetSalesByWarehouse(token, warehouseId) {
  try {
    const today = new Date();
    const fromDate = new Date(today);
    fromDate.setDate(today.getDate() - 30);

    const from = fromDate.toISOString().split('T')[0];
    const to = today.toISOString().split('T')[0];

    console.log(`\nTesting /sales with warehouse ${warehouseId} from ${from} to ${to}...`);
    const response = await fetch(`${PROHANDEL_CONFIG.API_URL}/sales?from=${from}&to=${to}&location_id=${warehouseId}`, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'X-Tenant-ID': PROHANDEL_CONFIG.TENANT_ID,
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch sales data: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    console.log('Sales data for warehouse:');
    console.log(JSON.stringify(data.slice(0, 2), null, 2)); // Show just the first 2 items
    console.log(`...and ${data.length - 2} more items`);
    return data;
  } catch (error) {
    console.error('Error fetching sales data for warehouse:', error.message);
    return null;
  }
}

// Run all tests
async function runTests() {
  console.log('Testing ProHandel API integration...');
  console.log('================================');
  
  const token = await authenticate();
  
  // Test warehouses endpoint
  const warehousesData = await testGetWarehouses(token);
  
  if (warehousesData && warehousesData.locations && warehousesData.locations.length > 0) {
    const firstWarehouseId = warehousesData.locations[0].id;
    
    // Test stockout risk with warehouse filter
    await testGetStockoutRiskByWarehouse(token, firstWarehouseId);
    
    // Test sales with warehouse filter
    await testGetSalesByWarehouse(token, firstWarehouseId);
  }
  
  console.log('\nAll tests completed!');
}

// Execute tests
runTests().catch(err => {
  console.error('Test execution failed:', err);
  process.exit(1);
});
